dask-gateway-server/dask_gateway_server/backends/kubernetes/controller.py

```diff
@@ -342,35 +380,45 @@ def get_cluster_key(self, obj):
 
             sched_pod_name = status.get("schedulerPod")
             if not sched_pod_name:
-                sched_pod_name = await self.create_scheduler_pod_if_not_exists(cluster)
+                sched_pod_name, sched_pod = await self.create_scheduler_pod_if_not_exists(
+                    cluster
+                )
                 status_update["schedulerPod"] = sched_pod_name
-
-            sched_pod = await self.read_namespaced_pod(sched_pod_name, namespace)
+            else:
+                sched_pod = await self.read_namespaced_pod(sched_pod_name, namespace)
 
             if sched_pod is not None:
                 sched_state = self.get_pod_state(sched_pod)
             else:
                 sched_state = "terminated"
             if sched_state == "running":
+                service_name = status.get("service")
                 if not status.get("service"):
                     service_name = await self.create_service_if_not_exists(
                         cluster, sched_pod
                     )
                     status_update["service"] = service_name
 
-                if not status.get("ingressroute"):
-                    route = await self.create_ingressroute_if_not_exists(
-                        cluster, sched_pod
-                    )
-                    status_update["ingressroute"] = route
-
-                if not status.get("ingressroutetcp"):
-                    route = await self.create_ingressroutetcp_if_not_exists(
-                        cluster, sched_pod
-                    )
-                    status_update["ingressroutetcp"] = route
-
-                status_update["phase"] = "Running"
+                endpoints = self.informers["endpoints"].get(
+                    f"{namespace}.{service_name}"
+                )
+                if endpoints is not None and self.endpoints_all_ready(endpoints):
+                    # Only add routes if endpoints exist, otherwise traefik gets grumpy
```

Comment:
Just out of curiosity, what do you mean by `grumpy`?

Reply:
It just logs a lot of errors, everything still works fine. We also want to delay moving the cluster into the running state until the route is likely to succeed being added to the proxy (unfortunately we can't be *sure* it's added unless we ping the proxy (which I don't want to do), but if the endpoint already exists, adding it to the proxy should happen almost immediately).